scripts/research/b2_01_volatility_transition_lib.py

```
@@ -0,0 +1,1033 @@
+
+def build_panel(frame_1m: Mapping[str, np.ndarray]) -> dict[str, np.ndarray]:
+    """Build only frozen B2-01 features/targets on the causal 15m grid."""
+    validate_1m_frame(frame_1m)
+    open_ms = np.asarray(frame_1m["open_time_ms"], dtype=np.int64)
+    avail = np.asarray(frame_1m["available_at_ms"], dtype=np.int64)
+    close = np.asarray(frame_1m["close"], dtype=np.float64)
+
+    grid = _grid_ms()
+    offset = ((grid - int(open_ms[0])) // BAR_MS).astype(np.int64)
+    if np.any(grid < open_ms[0]) or np.any(offset < 0):
+        raise B201InputError("decision grid precedes authorized 1m frame")
```

Ruff (0.16.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)